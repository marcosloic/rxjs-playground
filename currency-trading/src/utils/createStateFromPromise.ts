import {
    combineLatest,
    ConnectableObservable,
    from,
    Observable,
    of
} from 'rxjs';
import {
    catchError,
    map,
    publishReplay,
    startWith,
    switchMap,
    switchMapTo,
    take
} from 'rxjs/operators';

export type AsyncState<T> = {
    data: T | null;
    isLoading: boolean;
    error: string | null;
};

export const createStateFromPromise = <T>(
    request: (base: string, quote: string) => Promise<T>,
    firstArg$: Observable<string>,
    secondArg$: Observable<string>,
    loadIntent$: Observable<any>
): Observable<AsyncState<T>> => {
    const argumentStream$ = combineLatest([firstArg$, secondArg$]).pipe(
        take(1)
    );
    const sink$ = loadIntent$.pipe(
        switchMapTo(argumentStream$),
        switchMap(data => {
            const [a, b] = data;
            return from(request(a, b)).pipe(
                map((value: any) => {
                    return {
                        data: value,
                        isLoading: false,
                        error: null
                    };
                }),
                startWith({
                    data: null,
                    isLoading: true,
                    error: null
                }),
                catchError((error: string) => {
                    return of({
                        data: null,
                        isLoading: false,
                        error
                    });
                })
            );
        }),
        publishReplay(1)
    );
    (sink$ as ConnectableObservable<AsyncState<T>>).connect();
    return sink$;
};
