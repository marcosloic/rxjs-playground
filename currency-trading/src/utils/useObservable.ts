import { useState, useLayoutEffect } from 'react';
import { Observable } from 'rxjs';

export const useObservable = <T>(
    observable$: Observable<T>,
    initialValue?: any
): T => {
    const [value, update] = useState(initialValue);

    useLayoutEffect(() => {
        const s = observable$.subscribe(update);
        return () => s.unsubscribe();
    }, [observable$]);

    return value;
};
