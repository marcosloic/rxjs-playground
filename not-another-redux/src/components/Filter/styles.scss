.filter {
    padding: 10px;
    .title {
        text-transform: capitalize;
        margin-bottom: 5px;
    }
}
