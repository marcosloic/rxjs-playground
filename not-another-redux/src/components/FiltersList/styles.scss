.filtersList {
    display: flex;
    margin: 10px;
    flex-direction: column;

    .list {
        margin-bottom: 10px;
    }
}
