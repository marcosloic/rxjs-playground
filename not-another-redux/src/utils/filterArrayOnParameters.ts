import { ApiData } from '../types/ApiData.interface';

export const filterArrayOnParameters = (
    dataArray: ApiData[],
    parameters: { [k: string]: string }
): ApiData[] => {
    const parametersArray = Object.entries(parameters);

    return dataArray.filter(entry => {
        const isValid = parametersArray.reduce((acc, val) => {
            const [key, filterVal] = val;
            const target = entry[key];
            if (Array.isArray(target)) {
                return acc && target.indexOf(filterVal) !== -1;
            }
            return acc && entry[key] === filterVal;
        }, true);
        return isValid;
    });
};
