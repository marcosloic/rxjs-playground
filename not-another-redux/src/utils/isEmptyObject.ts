// Can be tripped up by Date objects without the constructor check
// https://stackoverflow.com/questions/679915/how-do-i-test-for-an-empty-javascript-object

export const isEmptyObject = obj =>
    Object.entries(obj).length === 0 && obj.constructor === Object;
