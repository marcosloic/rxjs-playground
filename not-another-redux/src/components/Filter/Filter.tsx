import React from 'react';
import { FiltersState } from '../../models/filters';
import './styles.scss';

type FilterPropTypes = {
    title: keyof FiltersState;
    values: FiltersState[keyof FiltersState];
    onChangeCallback: ({ title, value }) => void;
};

export const Filter: React.FunctionComponent<FilterPropTypes> = React.memo(
    props => {
        const { title, values, onChangeCallback } = props;
        const valuesArray = Array.from(values);
        const onValueChange = React.useCallback(
            event => {
                onChangeCallback({ title, value: event.target.value });
            },
            [title, onChangeCallback]
        );
        const defaultSelectValue = `Pick some ${title}`;
        return (
            <div className="filter">
                <h5 className={'title'}>{title}</h5>
                <select onChange={onValueChange} defaultValue={'default'}>
                    <option disabled value={'default'}>
                        {defaultSelectValue}
                    </option>
                    {valuesArray.sort().map(value => (
                        <option key={value} value={value}>
                            {value}
                        </option>
                    ))}
                </select>
            </div>
        );
    }
);

Filter.defaultProps = {
    title: '',
    values: new Set(),
    onChangeCallback: () => {}
};

Filter.displayName = 'FiltersList';
