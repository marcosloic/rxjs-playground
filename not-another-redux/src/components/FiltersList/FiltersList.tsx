import React from 'react';
import { FiltersState } from '../../models/filters';
import { Filter } from '../Filter/Filter';
import './styles.scss';
import {
    filterActionTypes,
    SelectedFilters,
    updateFiltersIntent$
} from '../../models/selectedFilters';
import { useObservable } from '../../utils/useObservable';
import { Observable } from 'rxjs';

type FiltersListPropTypes = {
    filters$: Observable<FiltersState>;
    selectedFilters$: Observable<SelectedFilters>;
};

export const FiltersList: React.FunctionComponent<
    FiltersListPropTypes
> = React.memo(props => {
    const { filters$, selectedFilters$ } = props;
    const onButtonClick = React.useCallback(() => {
        updateFiltersIntent$.next({ actionType: filterActionTypes.CLEAR });
    }, []);
    const onChangeCallback = React.useCallback(({ title, value }) => {
        updateFiltersIntent$.next({
            actionType: filterActionTypes.ADD,
            filter: title as string,
            value: value
        });
    }, []);

    const filters = useObservable(filters$, {});
    const selectedFilters = useObservable(selectedFilters$);
    const filtersArray = Object.entries(filters);
    const shouldDisplay = filtersArray && filtersArray.length > 0;
    const shouldDisplayButton =
        selectedFilters && Object.keys(selectedFilters).length > 0;
    if (!shouldDisplay) {
        return null;
    }
    return (
        <div className={'filtersList'}>
            <div className={'list'}>
                {filtersArray.map(filter => {
                    const [title, values] = filter;
                    return (
                        <Filter
                            key={title}
                            title={title}
                            values={values}
                            onChangeCallback={onChangeCallback}
                        />
                    );
                })}
            </div>
            {shouldDisplayButton && (
                <button className={'button'} onClick={onButtonClick}>
                    Reset filters
                </button>
            )}
        </div>
    );
});

FiltersList.displayName = 'FiltersList';
