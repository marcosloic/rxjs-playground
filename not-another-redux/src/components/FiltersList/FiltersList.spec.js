import {
    filterActionTypes,
    updateFiltersIntent$
} from '../../models/selectedFilters';

import TestRenderer from 'react-test-renderer';
import React from 'react';
import { FiltersList } from './FiltersList';
import { Filter } from '../Filter/Filter';
import { useObservable } from '../../utils/useObservable';

jest.mock('../../models/selectedFilters', () => {
    return {
        updateFiltersIntent$: {
            next: jest.fn()
        },
        filterActionTypes: {
            CLEAR: 'Clear',
            ADD: 'Add'
        }
    };
});

jest.mock('../../utils/useObservable');

const fakeFilters = {
    brand: 'brand',
    colors: 'white'
};

describe('The FiltersList component', () => {
    beforeEach(() => {
        useObservable.mockReturnValue(fakeFilters);
    });
    it('Matches the snapshot', () => {
        const filtersList = TestRenderer.create(<FiltersList />);
        expect(filtersList.toJSON()).toMatchSnapshot();
    });
    it('Renders select options for each filter', () => {
        const filtersList = TestRenderer.create(<FiltersList />);
        const instance = filtersList.root;
        const children = instance.findAllByType(Filter);
        expect(children.length).toBe(2);
    });
    it(`Sends a ${filterActionTypes.ADD} when the Filter emits a value`, () => {
        // Roughly tests that the filter works as well
        const filtersList = TestRenderer.create(<FiltersList />);
        const instance = filtersList.root;
        const children = instance.findAllByType(Filter);
        const selectElement = children[0].findByProps({
            defaultValue: 'default'
        });
        selectElement.props.onChange({ target: { value: 'value' } });
    });

    it(`Sends a ${filterActionTypes.CLEAR} when the user clicks on the button`, () => {
        const filtersList = TestRenderer.create(<FiltersList />);
        const instance = filtersList.root;
        const button = instance.findByProps({ className: 'button' });
        button.props.onClick();
        expect(updateFiltersIntent$.next).toHaveBeenCalledWith({
            actionType: 'Clear'
        });
    });
});
