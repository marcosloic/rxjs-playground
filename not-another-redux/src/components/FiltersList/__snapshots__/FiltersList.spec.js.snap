// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The FiltersList component Matches the snapshot 1`] = `
<div
  className="filtersList"
>
  <div
    className="list"
  >
    <div
      className="filter"
    >
      <h5
        className="title"
      >
        brand
      </h5>
      <select
        defaultValue="default"
        onChange={[Function]}
      >
        <option
          disabled={true}
          value="default"
        >
          Pick some brand
        </option>
        <option
          value="a"
        >
          a
        </option>
        <option
          value="b"
        >
          b
        </option>
        <option
          value="d"
        >
          d
        </option>
        <option
          value="n"
        >
          n
        </option>
        <option
          value="r"
        >
          r
        </option>
      </select>
    </div>
    <div
      className="filter"
    >
      <h5
        className="title"
      >
        colors
      </h5>
      <select
        defaultValue="default"
        onChange={[Function]}
      >
        <option
          disabled={true}
          value="default"
        >
          Pick some colors
        </option>
        <option
          value="e"
        >
          e
        </option>
        <option
          value="h"
        >
          h
        </option>
        <option
          value="i"
        >
          i
        </option>
        <option
          value="t"
        >
          t
        </option>
        <option
          value="w"
        >
          w
        </option>
      </select>
    </div>
  </div>
  <button
    className="button"
    onClick={[Function]}
  >
    Reset filters
  </button>
</div>
`;
